File: src/tui/slash/catalog.rs


#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum SlashKind {

    Instant,

    Toggle,

    Skill,

    Anchor,

    Panel(PanelKind),

    Auth,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum SettingsTab {
    Status,
    Config,
    Usage,
}

impl SettingsTab {

    pub fn slash_name(self) -> &'static str {
        match self {
            SettingsTab::Status => "status",
            SettingsTab::Config => "config",
            SettingsTab::Usage => "usage",
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum PanelKind {
    Help,
    Resume,
    Rewind,
    Settings(SettingsTab),
    Model,
    Effort,
    Permissions,
    Hooks,
    Diff,
    Skills,
    Agents,
    Mcp,

    Tasks,
}

impl PanelKind {

    pub fn slash_name(self) -> &'static str {
        match self {
            PanelKind::Help => "help",
            PanelKind::Resume => "resume",
            PanelKind::Rewind => "rewind",
            PanelKind::Settings(tab) => tab.slash_name(),
            PanelKind::Model => "model",
            PanelKind::Effort => "effort",
            PanelKind::Permissions => "permissions",
            PanelKind::Hooks => "hooks",
            PanelKind::Diff => "diff",
            PanelKind::Skills => "skills",
            PanelKind::Agents => "agents",
            PanelKind::Mcp => "mcp",
            PanelKind::Tasks => "tasks",
        }
    }
}

#[derive(Debug, Clone, Copy)]
pub struct SlashEntry {

    pub name: &'static str,

    pub brief: &'static str,

    pub kind: SlashKind,
}

pub const CATALOG: &[SlashEntry] = &[

    SlashEntry {
        name: "exit",
        brief: "exit the TUI",
        kind: SlashKind::Instant,
    },
    SlashEntry {
        name: "clear",
        brief: "wipe history, re-splash mascot",
        kind: SlashKind::Instant,
    },

    SlashEntry {
        name: "plan",
        brief: "enable plan mode",
        kind: SlashKind::Toggle,
    },
    SlashEntry {
        name: "copy",
        brief: "copy the last assistant response to clipboard",
        kind: SlashKind::Toggle,
    },
    SlashEntry {
        name: "export",
        brief: "export the current conversation to a file or clipboard",
        kind: SlashKind::Toggle,
    },

    SlashEntry {
        name: "dream",
        brief: "reflective memory consolidation",
        kind: SlashKind::Skill,
    },
    SlashEntry {
        name: "review",
        brief: "code review a pull request",
        kind: SlashKind::Skill,
    },
    SlashEntry {
        name: "init",
        brief: "initialize project OTHERSIDE.md",
        kind: SlashKind::Skill,
    },
    SlashEntry {
        name: "security-review",
        brief: "review pending changes for security issues",
        kind: SlashKind::Skill,
    },
    SlashEntry {
        name: "loop",
        brief: "run a prompt or slash on a recurring interval",
        kind: SlashKind::Skill,
    },

    SlashEntry {
        name: "branch",
        brief: "fork the conversation from here",
        kind: SlashKind::Anchor,
    },
    SlashEntry {
        name: "compact",
        brief: "summarize history, trim tokens",
        kind: SlashKind::Anchor,
    },
    SlashEntry {
        name: "context",
        brief: "visualize current context usage",
        kind: SlashKind::Anchor,
    },

    SlashEntry {
        name: "help",
        brief: "show slash command catalog",
        kind: SlashKind::Panel(PanelKind::Help),
    },
    SlashEntry {
        name: "resume",
        brief: "pick a past session to continue",
        kind: SlashKind::Panel(PanelKind::Resume),
    },
    SlashEntry {
        name: "rewind",
        brief: "jump back to an earlier turn",
        kind: SlashKind::Panel(PanelKind::Rewind),
    },
    SlashEntry {
        name: "config",
        brief: "show the config file path",
        kind: SlashKind::Panel(PanelKind::Settings(SettingsTab::Config)),
    },
    SlashEntry {
        name: "model",
        brief: "show or switch the active model",
        kind: SlashKind::Panel(PanelKind::Model),
    },
    SlashEntry {
        name: "effort",
        brief: "tune reasoning effort (low/med/high/max/auto)",
        kind: SlashKind::Panel(PanelKind::Effort),
    },
    SlashEntry {
        name: "permissions",
        brief: "manage tool permission rules",
        kind: SlashKind::Panel(PanelKind::Permissions),
    },
    SlashEntry {
        name: "hooks",
        brief: "view or edit hooks for tool events",
        kind: SlashKind::Panel(PanelKind::Hooks),
    },
    SlashEntry {
        name: "diff",
        brief: "view uncommitted or per-turn diffs",
        kind: SlashKind::Panel(PanelKind::Diff),
    },
    SlashEntry {
        name: "skills",
        brief: "list available skills",
        kind: SlashKind::Panel(PanelKind::Skills),
    },
    SlashEntry {
        name: "agents",
        brief: "manage agent configurations",
        kind: SlashKind::Panel(PanelKind::Agents),
    },
    SlashEntry {
        name: "status",
        brief: "show session status (Status tab of Settings)",
        kind: SlashKind::Panel(PanelKind::Settings(SettingsTab::Status)),
    },
    SlashEntry {
        name: "usage",
        brief: "show plan usage limits (Usage tab of Settings)",
        kind: SlashKind::Panel(PanelKind::Settings(SettingsTab::Usage)),
    },
    SlashEntry {
        name: "mcp",
        brief: "manage MCP servers",
        kind: SlashKind::Panel(PanelKind::Mcp),
    },
    SlashEntry {
        name: "tasks",
        brief: "list and manage background tasks",
        kind: SlashKind::Panel(PanelKind::Tasks),
    },
    SlashEntry {
        name: "bashes",
        brief: "list and manage background tasks",
        kind: SlashKind::Panel(PanelKind::Tasks),
    },

    SlashEntry {
        name: "login",
        brief: "sign in to a provider",
        kind: SlashKind::Auth,
    },
    SlashEntry {
        name: "logout",
        brief: "sign out from a provider",
        kind: SlashKind::Auth,
    },
];

pub fn lookup(name: &str) -> Option<&'static SlashEntry> {
    CATALOG.iter().find(|e| e.name.eq_ignore_ascii_case(name))
}

pub fn prefix_matches(prefix_lower: &str) -> impl Iterator<Item = &'static SlashEntry> + use<'_> {
    CATALOG
        .iter()
        .filter(move |e| e.name.to_ascii_lowercase().starts_with(prefix_lower))
}

pub fn display_line(entry: &SlashEntry) -> String {
    format!("/{} — {}", entry.name, entry.brief)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn lookup_case_insensitive() {
        assert!(lookup("HELP").is_some());
        assert!(lookup("Clear").is_some());
    }

    #[test]
    fn no_leading_slash_in_names() {
        for e in CATALOG {
            assert!(
                !e.name.starts_with('/'),
                "name must not include leading slash: {}",
                e.name
            );
        }
    }

    #[test]
    fn names_are_ascii_lowercase() {
        for e in CATALOG {
            assert!(
                e.name.chars().all(|c| c.is_ascii_lowercase() || c == '-'),
                "non-lowercase or non-ascii name: {}",
                e.name
            );
        }
    }

    #[test]
    fn names_are_unique() {
        let mut seen = std::collections::HashSet::new();
        for e in CATALOG {
            assert!(seen.insert(e.name), "duplicate slash name: {}", e.name);
        }
    }

    #[test]
    fn briefs_non_empty() {
        for e in CATALOG {
            assert!(!e.brief.is_empty(), "brief missing for /{}", e.name);
        }
    }

    #[test]
    fn prefix_matches_finds_s_group() {

        let names: Vec<&str> = prefix_matches("s").map(|e| e.name).collect();
        assert!(names.contains(&"status"));
        assert!(names.contains(&"skills"));
        assert!(names.contains(&"security-review"));
        assert!(!names.contains(&"swarm"), "/swarm cut in 011");
        assert!(!names.contains(&"statusline"), "/statusline cut — otherside has no statusLine feature");
    }

    #[test]
    fn display_line_format() {
        let e = lookup("help").unwrap();
        assert_eq!(display_line(e), "/help — show slash command catalog");
    }

    #[test]
    fn panel_kind_slash_names_match_catalog() {

        const ALIAS_NAMES: &[&str] = &["bashes"];
        for entry in CATALOG {
            if ALIAS_NAMES.contains(&entry.name) {
                continue;
            }
            if let SlashKind::Panel(kind) = entry.kind {
                assert_eq!(
                    entry.name,
                    kind.slash_name(),
                    "PanelKind::{:?} drift vs /{}",
                    kind,
                    entry.name
                );
            }
        }
    }

}

File: src/commands/catalog.ts
import { get as getSkill } from "@/engine/skills/registry.ts";
import { commandHint } from "./hints.ts";

export type SlashKind =
  | "instant"
  | "toggle"
  | "skill"
  | "workflow"
  | "anchor"
  | "panel"
  | "auth"
  | "external";

export interface SlashCommand {
  name: string;
  kind: SlashKind;
  description: string;
  argumentHint?: string;
  aliases?: readonly string[];
}

const EARLY_SKILL_ORDER = ["dream", "pr-review", "init", "loop"] as const;
const LATE_SKILL_ORDER = ["pr-security-review", "deep-security-review", "grill-me"] as const;

function skillCommand(name: string): SlashCommand | null {
  const skill = getSkill(name);
  if (!skill?.userInvocable) return null;
  return {
    name: skill.name,
    kind: "skill",
    description: commandHint(skill.name, skill.description || "Dynamic skill"),
    ...(skill.argumentHint ? { argumentHint: skill.argumentHint } : {}),
    ...(skill.aliases.length > 0 ? { aliases: skill.aliases } : {}),
  };
}

function skillCommands(names: readonly string[]): SlashCommand[] {
  return names.flatMap((name) => {
    const command = skillCommand(name);
    return command ? [command] : [];
  });
}

export function buildCatalog(): SlashCommand[] {
  const commands: SlashCommand[] = [
    { name: "exit", kind: "instant", description: "exit the TUI", aliases: ["quit"] },
    {
      name: "clear",
      kind: "instant",
      description: "wipe history, re-splash mascot",
    },
    {
      name: "cd",
      kind: "instant",
      description: "Change the current working directory",
      argumentHint: "<path>",
    },
    {
      name: "plan",
      kind: "toggle",
      description: "enable plan mode",
      argumentHint: "[<description>]",
    },
    {
      name: "fast",
      kind: "toggle",
      description: "toggle fast mode - codex only",
      argumentHint: "[on|off]",
    },
    {
      name: "parallel",
      kind: "toggle",
      description: "toggle parallel agent tasks",
      argumentHint: "[on|off]",
    },
    {
      name: "multiprovider",
      kind: "toggle",
      description: "toggle multi-provider routing",
      argumentHint: "[on|off]",
    },
    {
      name: "copy",
      kind: "toggle",
      description: "copy the last assistant response (/copy N for older)",
      argumentHint: "[N]",
    },
    {
      name: "export",
      kind: "toggle",
      description: "export the current conversation to a file or clipboard",
      argumentHint: "[file|clipboard]",
    },
    {
      name: "toggle-memory",
      kind: "toggle",
      description: "toggle auto memory for this session",
    },
    ...skillCommands(EARLY_SKILL_ORDER),
    {
      name: "btw",
      kind: "instant",
      description: "answer a side question without disturbing the main thread",
      argumentHint: "<question>",
      aliases: ["sidequest"],
    },
    ...skillCommands(LATE_SKILL_ORDER),
    {
      name: "ultraplan",
      kind: "workflow",
      description: "deep multi-agent implementation plan — explore, design, critique, synthesize",
      argumentHint: "<what to plan>",
    },
    {
      name: "goal",
      kind: "anchor",
      description: "Set a goal — keep working until the condition is met",
      argumentHint: "[<condition> | clear]",
    },
    {
      name: "branch",
      kind: "anchor",
      description: "fork the conversation from here",
      argumentHint: "[name]",
    },
    {
      name: "fork",
      kind: "instant",
      description: "Spawn a background agent that inherits the full conversation",
      argumentHint: "<directive>",
    },
    {
      name: "compact",
      kind: "anchor",
      description: "summarize history, trim tokens",
      argumentHint: "[custom summarization instructions]",
    },
    {
      name: "context",
      kind: "anchor",
      description: "visualize current context usage",
    },
    { name: "help", kind: "panel", description: "show slash command catalog" },
    {
      name: "resume",
      kind: "panel",
      description: "pick a past session to continue",
      argumentHint: "[session id or search]",
    },
    {
      name: "rewind",
      kind: "panel",
      description: "restore the code and/or conversation to a previous point",
      aliases: ["checkpoint"],
    },
    {
      name: "config",
      kind: "panel",
      description: "show the config file path",
      argumentHint: "[details]",
    },
    {
      name: "model",
      kind: "panel",
      description: "show or switch the active model",
      argumentHint: "[model]",
    },
    {
      name: "theme",
      kind: "panel",
      description: "show or switch the active theme",
    },
    {
      name: "effort",
      kind: "panel",
      description: "tune reasoning effort (low/med/high/xhigh/max/auto)",
      argumentHint: "[low|medium|high|xhigh|max|auto]",
    },
    {
      name: "permissions",
      kind: "panel",
      description: "manage tool permission rules",
    },
    {
      name: "hooks",
      kind: "panel",
      description: "view or edit hooks for tool events",
    },
    {
      name: "diff",
      kind: "panel",
      description: "view uncommitted or per-turn diffs",
    },
    { name: "skills", kind: "panel", description: "list available skills" },
    { name: "agents", kind: "panel", description: "manage agent configurations" },
    { name: "status", kind: "panel", description: "show session status" },
    { name: "usage", kind: "panel", description: "show provider usage" },
    { name: "stats", kind: "panel", description: "show provider stats" },
    {
      name: "mcp",
      kind: "panel",
      description: "manage MCP servers",
      argumentHint: "[enable|disable [server-name]]",
    },
    {
      name: "plugins",
      kind: "panel",
      aliases: ["plugin"],
      description: "manage plugins and marketplaces",
      argumentHint: "[install|update|uninstall|enable|disable <plugin@marketplace>|list]",
    },
    {
      name: "marketplace",
      kind: "instant",
      description: "manage plugin marketplaces",
      argumentHint: "[add <source>|remove <name>|update [name]|list]",
    },
    {
      name: "reload",
      kind: "instant",
      aliases: ["reload-plugins"],
      description: "reload skills, agents, and MCP servers without restarting",
    },
    {
      name: "tasks",
      kind: "panel",
      description: "list and manage background tasks",
    },
    {
      name: "bashes",
      kind: "panel",
      description: "list and manage background shells",
    },
    {
      name: "workflows",
      kind: "panel",
      description: "list and manage background workflows",
    },
    {
      name: "remote",
      kind: "panel",
      description: "link a phone and manage remote sessions",
    },
    {
      name: "login",
      kind: "auth",
      description: "sign in to a provider",
      argumentHint: "[provider]",
    },
    {
      name: "logout",
      kind: "auth",
      description: "sign out from a provider",
      argumentHint: "[provider]",
    },
    {
      name: "design",
      kind: "panel",
      description: "manage Otherside Design sessions",
      argumentHint: "[prompt]",
    },
  ];
  return commands.map((command) => ({
    ...command,
    description: commandHint(command.name, command.description),
  }));
}

export const CATALOG: SlashCommand[] = new Proxy([] as SlashCommand[], {
  get(_target, property) {
    const catalog = buildCatalog();
    const value = Reflect.get(catalog, property, catalog);
    return typeof value === "function" ? value.bind(catalog) : value;
  },
});
